exonum/src/runtime/mod.rs

```diff
@@ -120,53 +122,118 @@ impl From<RuntimeIdentifier> for u32 {
         spec: Any,
     ) -> Box<dyn Future<Item = (), Error = ExecutionError>>;
 
-    /// Returns additional information about artifact with the specified id if it is deployed.
+    /// Returns protobuf description of deployed artifact with the specified identifier.
     fn artifact_info(&self, id: &ArtifactId) -> Option<ArtifactInfo>;
 
     /// Starts a new service instance with the given specification.
+    ///
+    /// # Policy on panics
+    ///
+    /// * This method should catch each kind of panics except of `FatalError` and converts
+    /// them into `ExecutionError`.
+    /// * If panic occurs, the runtime must ensure that it is in a consistent state.
     fn start_service(&mut self, spec: &InstanceSpec) -> Result<(), ExecutionError>;
 
     /// Configures a service instance with the given parameters.
+    ///
+    /// There are two cases when this method is called:
+    ///
+    /// - After creating a new service instance by the [`start_service`] invocation, in this case
+    /// if an error during this action occurs, dispatcher will invoke [`stop_service`]
+    /// and you must be sure that this invocation will not fail.
+    /// - During the configuration change procedure. [ECR-3306]
+    ///
+    /// # Policy on panics
+    ///
+    /// * This method should catch each kind of panics except of `FatalError` and converts
+    /// them into `ExecutionError`.
+    ///
+    /// ['start_service`]: #start_service
+    /// ['stop_service`]: #stop_service
     fn configure_service(
         &self,
-        context: &Fork,
+        fork: &Fork,
         spec: &InstanceSpec,
         parameters: Any,
     ) -> Result<(), ExecutionError>;
 
     /// Stops existing service instance with the given specification.
+    ///
+    /// # Policy on panics
+    ///
+    /// * This method should catch each kind of panics except of `FatalError` and converts
+    /// them into `ExecutionError`.
+    /// * If panic occurs, the runtime must ensure that it is in a consistent state.
     fn stop_service(&mut self, spec: &InstanceSpec) -> Result<(), ExecutionError>;
 
-    /// Execute transaction.
-    // TODO Do not use dispatcher struct directly.
+    /// Execute service transaction.
+    ///
+    /// # Policy on panics
+    ///
+    /// Do not process, just skip above.
     fn execute(
         &self,
         dispatcher: &dispatcher::Dispatcher,
         context: &mut ExecutionContext,
         call_info: CallInfo,
-        payload: &[u8],
+        arguments: &[u8],
     ) -> Result<(), ExecutionError>;
 
     /// Gets state hashes of the every contained service.
     fn state_hashes(&self, snapshot: &dyn Snapshot) -> StateHashAggregator;
 
     /// Calls `before_commit` for all the services stored in the runtime.
+    ///
+    /// # Policy on panics
+    ///
+    /// * This method should catch each kind of panics except of `FatalError` and writes
+    /// them into log.
+    /// * If panic occurs, the runtime should rollback changes in fork.
```

Comment:
Shall it rollback all changes in the fork, or made by the service that caused the panic?

Reply:
Just only have to revert changes from service that caused panic.